docs/protocol-generator.md

```diff
@@ -72,30 +82,52 @@ The specific types which could be assigned to contents in a protocol specificati
+| Primitive collection types | `<PCT>` | `pt:set[<PT>]`                                                                                          | `pt:set[pt:str]`                         | `FrozenSet[str]`                   |
+|                            |               | `pt:list[<PT>]`                                                                                         | `pt:list[pt:int]`                        | `Tuple[int, ...]`<sup>*</sup>      |
+| Primitive mapping types<sup>2</sup>    | `<PMT>` | `pt:dict[<PT>, <PT>]`                                                                 | `pt:dict[pt:str, pt:bool]`               | `Dict[str, bool]`                  |
+| Multi types                | `<MT>`  | `pt:union[<PT>/<CT>/<PCT>/<PMT>, ..., <PT>/<CT>/<PCT>/<PMT>]` | `pt:union[ct:DataModel, pt:set[pt:str]]` | `Union[DataModel, FrozenSet[str]]` |
+| Optional types             | `<O>`   | `pt:optional[<MT>/<PMT>/<PCT>/<PT>/<CT>]`                                       | `pt:optional[pt:bool]`                   | `Optional[bool]`                   |
 
 <sup>*</sup> <sub>This is how variable length tuples containing elements of the same type are declared in Python; see <a href="https://docs.python.org/3/library/typing.html#typing.Tuple" target=_blank>here</a> </sub>
 
+#### Notes
+
+1. Currently, there is no way to describe custom types in a language independent format. This means that if a protocol specification includes custom types, the required implementations must be provided manally.
+    * Before generating the protocol, the protocol buffer schema code for every custom type must be provided in the protocol specification. You can see an example of this in the above protocol specification. In particular, if a specification includes custom types, the specification must include a second yaml document (each yaml document is enclosed within `---` and `...`) which for every custom type describes its protocol buffer schema code.   
+    * Once the generator is called, it produces a `custom_types` module containing stub implmentations for every custom type in the specification. The user must then modify this module and add implemntations for every custom type in the specification. This includes implementations of how an object of a custom type can be encoded and decoded using protocol buffer.
+    * Note, currently the way custom types are dealt with in the generator is admittedly inconvinient. The reason is, the generator does not know the structure of custom types and how they may be serialised/deserialised. Although this approach works, it is only a temporary solution until further work on an infrastructure-independent type description language is finished (similar to how the generator is designed to be an infrastructure-independent protocol description language).
```

Comment:
I would not call it infrastructure-independent..

Reply:
Done